mlflow/utils/__init__.py

```diff
@@ -120,3 +119,37 @@ def _get_fully_qualified_class_name(obj):
     Obtains the fully qualified class name of the given object.
     """
     return obj.__class__.__module__ + "." + obj.__class__.__name__
+
+
+def _inspect_original_var_name(var):
+    import inspect
+    try:
+        original_var_name = None
+
+        frame = inspect.currentframe().f_back
+        while frame is not None:
+            arg_info = inspect.getargvalues(frame)
+
+            fixed_args = [arg_info.locals[arg_name] for arg_name in arg_info.args]
+            varlen_args = list(arg_info.locals[arg_info.varargs]) if arg_info.varargs else []
+            keyword_args = list(arg_info.locals[arg_info.keywords].values()) if arg_info.keywords else []
+
+            all_args = fixed_args + varlen_args + keyword_args
+
+            # check whether `var` is in arg list first. If yes, go to check parent frame.
+            if any(var is arg for arg in all_args):
+                # the var is passed in from caller, check parent frame.
+                frame = frame.f_back
+                continue
+
+            for var_name, var_val in frame.f_locals.items():
+                if var_val is var:
+                    original_var_name = var_name
+                    break
+
+            break
+
+        return original_var_name
+
+    except BaseException:
+        pass
```

Comment:
This method will return the inspected the original var name, by checking ancestor frames local_vars.

We can test it by:
```
from mlflow.utils import _inspect_original_var_name


def f1(a, *args, **kwargs):
  v0 = a
  print(f'arg-a is {_inspect_original_var_name(v0)}')
  print(f'args[0] is {_inspect_original_var_name(args[0])}')
  print(f"kwargs v1 is {_inspect_original_var_name(kwargs['v1'])}")

def f2(v, *k1, **kw1):
  f1(v, *k1, **kw1)

ta1=1
ta2=2
ta3=3
ta4=4

f2(ta1, ta2, v1=ta3, v2=ta4)
```
We can get output:
```
arg-a is ta1
args[0] is ta2
kwargs v1 is ta3
```

Reply:
What would happen if `f2` looked like:

```
def f2(v, *k1, **kw1):
  v_processed = func1(v)
  f1(v_processed, *k1, **kw1)
```

Reply:
This case we cannot capture the `v` but we will capture name `v_processed`

Reply:
In doc we can say: the auto capturing dataset name, we can only capture the variable name which is exactly the same instance with the one fetched in metric API.

Reply:
A TODO:
Suppose the case:

```
eval1_X, eval1_y = load_data1()
eval2_X, eval2_y = load_data2()

def compute_metric(X, y):
   processed_X, processed_y = process_eval_data(X, y)
   return model.score(processed_X, processed_y)

compute_metric(eval1_X, eval1_y)
compute_metric(eval2_X, eval2_y)
```

This case the autolog metric will log metric name as "processed_X" for both eval1 and eval2 data. And the second logged metric will **overwrite** the first logged metric value. To avoid overwriting, we should detect this case and log the second metric name as `processed_X_2`

Reply:
Sounds like a good follow-up. Can we add an inline comment in an appropriate location within the code with a TODO for this?